### DISCUSSION TITLE
Why Rust?

### DISCUSSION THREAD

**Discussion by marpit19** [Cormanz/smartgpt#46] [Q&A]
I am right now learning more things about rust so I wanted to ask why you implemented this using Rust instead of using Python or Golang

**Comment by Cormanz** [comment]
Hi! I decided to implement this in Rust for a couple of reasons.

1. Rust is one of the fastest languages out there, which means if we want to integrate with tools that are computationally expensive, we could do that very quickly. In particular, it has no garbage collector, unlike Go (as far as I know.)
2. Rust has very strong ergonomics for a systems language, with traits, lambdas, and etc. keeping code rather readable.
3. One of the most important things is that Rust allows for the codebase to be very easy to maintain. We can always ensure that our codebase is memory-safe because that's a Rust guarantee (and if we introduce unsafe code, we can easily isolate it), and Rust has a very strong static typing system that catches bugs. A language like Python doesn't have this.
4. In the future, we plan to make it so you can just use SmartGPT from one executable binary. Rust makes it super easy to build this, unlike Python. And just in general, using SmartGPT is incredible easy as cargo automatically installs and builds dependencies.
5. Rust has a large ecosystem of external crates we can take advantage of, including crates for support with large language models (i.e. `llm-rs`,.)
6. I initially made this project in Rust to help me grow more experienced in the language as well.

In general, I find the language to be the sweet-spot between performance, maintainability, and ergonomics for this project.
